### PR TITLE
[Feature] Enabled un-seen clients case to check the participation generalization gap

### PR DESCRIPTION
**Pull request by yxdyc**
Enabled un-seen clients simulation, such that we can check the participation generalization gap, refer [ICLR'22, What Do We Mean by Generalization in Federated Learning?](https://openreview.net/forum?id=VimqQq-i_Q)

This PR is based on the another PR (enhanced monitor #90)

### CODE REVIEW

File: federatedscope/core/fed_runner.py [hunk 1]
```diff
@@ -89,7 +91,18 @@ def _setup_for_standalone(self):
...
+            unseen_clients_id = np.random.choice(
+                np.arange(1, self.cfg.federate.client_num + 1),
+                size=max(
+                    1,
+                    int(self.cfg.federate.unseen_clients_rate *
+                        self.cfg.federate.client_num)),
+                replace=False).tolist()
+
         self.server = self._setup_server()
+        self.server.unseen_clients_id = unseen_clients_id
```
Comment (xieyxclack) [inline federatedscope/core/fed_runner.py:105]: IMO, we have better not modify the attributes of server/client in fed_runner

File: federatedscope/core/fed_runner.py [hunk 2]
```diff
@@ -102,6 +115,8 @@ def _setup_for_standalone(self):
         for client_id in range(1, self.cfg.federate.client_num + 1):
             self.client[client_id] = self._setup_client(
                 client_id=client_id, client_model=self._shared_client_model)
+            if client_id in unseen_clients_id:
+                self.client[client_id].is_unseen_client = True
```
Comment (xieyxclack) [inline federatedscope/core/fed_runner.py:119]: Like above

File: federatedscope/core/sampler.py [hunk 3]
```diff
@@ -23,14 +23,19 @@ def change_state(self, indices, state):
         """
         To modify the state of clients (idle or working)
         """
-        if state == 'idle':
-            self.client_state[indices] = 1
-        elif state == 'working':
-            self.client_state[indices] = 0
+        if not isinstance(indices, list):
```
Comment (xieyxclack) [inline federatedscope/core/sampler.py:26]: Shall we cover np.array here?

**Review by xieyxclack — COMMENTED** [review]
LGTM, please see the inline comments for minor suggestions. And I am not sure that whether setting `unseen_clients` would affect the behaviors of broadcasting to all neighbors in some applications.

File: federatedscope/core/fed_runner.py (hunk 1 above)
Comment (yxdyc) [reply federatedscope/core/fed_runner.py]: moved it into the setup_server() as sub action of init

File: federatedscope/core/fed_runner.py (hunk 2 above)
Comment (yxdyc) [reply federatedscope/core/fed_runner.py]: moved it into the setup_server() as sub action of init

File: federatedscope/core/sampler.py (hunk 3 above)
Comment (yxdyc) [reply federatedscope/core/sampler.py]: modified accordingly, besides, we used another state named "unseen"

**Review by xieyxclack — APPROVED** [review]
LGTM

**Pull request merged**